components/core/src/os/signals/unix.rs

```diff
@@ -62,20 +52,122 @@ pub fn check_for_signal() -> Option<SignalEvent> {
     }
 }
 
-fn set_signal_handlers() {
-    unsafe {
-        signal(libc::SIGINT, handle_shutdown_signal);
-        signal(libc::SIGTERM, handle_shutdown_signal);
+fn start_signal_handler() -> io::Result<()> {
+    let mut handled_signals = Sigset::empty()?;
+    handled_signals.addsig(libc::SIGINT)?;
+    handled_signals.addsig(libc::SIGTERM)?;
+    handled_signals.addsig(libc::SIGHUP)?;
+    handled_signals.addsig(libc::SIGQUIT)?;
+    handled_signals.addsig(libc::SIGALRM)?;
+    handled_signals.addsig(libc::SIGUSR1)?;
+    handled_signals.addsig(libc::SIGUSR2)?;
+    handled_signals.addsig(libc::SIGCHLD)?;
+    handled_signals.block()?;
+    thread::Builder::new().name("signal-handler".to_string())
+                          .spawn(move || {
+                              loop {
+                                  // Using expect here seems reasonable because our understanding of wait() is that it only
+                                  // returns an error if we called it with a bad signal.
+                                  let signal = handled_signals.wait().expect("sigwait failed");
+                                  debug!("signal-handler thread received signal {:?}!", signal);
+                                  match signal {
+                                      libc::SIGINT | libc::SIGTERM => {
+                                          super::SHUTDOWN.store(true, Ordering::SeqCst);
+                                      }
+                                      _ => {
+                                          CAUGHT_SIGNALS.lock()
+                                                        .expect("Signal mutex poisoned")
+                                                        .push_back(signal);
+                                      }
+                                  };
+                              }
+                          })?;
+    Ok(())
+}
+
+// Sigset is a wrapper for the underlying libc type.
+struct Sigset {
+    inner: libc::sigset_t
+}
+
+impl Sigset {
```

Comment:
Is it worth looking at the nix crate (https://github.com/nix-rust/nix/blob/01f4d57eede3ed2f1051778113b1d6a7dff8e9d6/src/sys/signal.rs)
Looks like we already pull it in indirectly, but not directly.

Reply:
Happy to change it over if y'all prefer.  It seems the main difference in implementation (aside from them creating more types given their larger API surface area) is they use `mem::MaybeUninit::uninit()` whereas I use `mem::zeroed()` and `libc::c_int`.  Based on reading this: 

https://doc.rust-lang.org/beta/std/mem/union.MaybeUninit.html

It seems like that is a worthwhile change even if we don't move to the library.

Reply:
Although the documentation here makes me think that I've used mem::zeroed() as intended:

https://doc.rust-lang.org/std/mem/fn.zeroed.html

Reply:
I looked at the mem::zeroed() and mem::MaybeUninit::uninit() docs, and think that your mem::zeroed usage is correct.